src/solution.c

```diff
@@ -285,40 +295,35 @@ void send_observations(u8 n, gps_time_t *t, navigation_measurement_t *m)
   }
 }
 
-static BinarySemaphore solution_wakeup_sem;
-#define tim5_isr Vector108
-#define NVIC_TIM5_IRQ 50
-void tim5_isr()
+static BSEMAPHORE_DECL(solution_wakeup_sem, TRUE);
+static void sol_gpt_cb(GPTDriver *gptd)
 {
-  CH_IRQ_PROLOGUE();
-  chSysLockFromIsr();
+  (void)gptd;
+  chSysLockFromISR();
 
   /* Wake up processing thread */
   chBSemSignalI(&solution_wakeup_sem);
 
-  timer_clear_flag(TIM5, TIM_SR_UIF);
-
-  chSysUnlockFromIsr();
-  CH_IRQ_EPILOGUE();
+  chSysUnlockFromISR();
 }
 
-static void timer_set_period_check(uint32_t timer_peripheral, uint32_t period)
+static void timer_set_period_check(GPTDriver *gptd, uint32_t period)
 {
-  __asm__("CPSID i;");
-  TIM_ARR(timer_peripheral) = period;
-  uint32_t tmp = TIM_CNT(timer_peripheral);
+  chSysLock();
+  gptChangeIntervalI(gptd, period);
+  u32 tmp = gptGetCounterX(gptd);
   if (tmp > period) {
-    TIM_CNT(timer_peripheral) = period;
+    gpt_lld_set_counter(gptd, period);
```

Comment:
That's exactly why my Zynq GPT driver does this:

```void gpt_lld_change_interval(GPTDriver *gptp, gptcnt_t interval) {

  /* If the new interval value happens to be less than the current counter
   * value, then the counter will not reset until it overflows. To prevent
   * this, simply stop and restart the timer with the new interval.
   */

  gpt_lld_stop_timer(gptp);
  gpt_lld_start_timer(gptp, interval);
}```